### ISSUE TITLE
Gutenboarding: signup data should persist past login boundary

### ISSUE THREAD

**Issue by andrewserong** [Automattic/wp-calypso#42035]
In Calypso, when we log a user in, we call `user.clear()` just as we do when we log a user out. The intention of calling this is to ensure that we ["discard any user reference that the application may hold"](https://github.com/automattic/wp-calypso/blob/4537033c6794449ce0a10b69d72daf5fa658d553/client/lib/user/user.js#L231).

However, in Gutenboarding, where we have the signup form at the end of the flow, we wish newly logged in users to have access to the data they entered before logging in. In efforts such as https://github.com/Automattic/wp-calypso/issues/41771 and https://github.com/Automattic/wp-calypso/pull/41780 we attempted to return a logged in user to the point they left off, and this was working correctly in our local development environments, but appears not to work in production.

My hunch is that the issue was obfuscated for us while testing the magic login links, because when we click the links in the email, they redirect us to finish logging in at `wordpress.com` and then redirect us to `calypso.localhost:3000`. When we click that button and log in, Calypso calls `user.clear()` and clears out the local storage, but because it's only cleared at wordpress.com, our environment at `calypso.localhost:3000` is blessedly protected from the destructive action, and it looked like everything was working correctly.

To get this sort of data persistence working correctly past the login boundary, it looks like we'll need to handle it in the `rebootAfterLogin` method in `handle-emailed-link-form.jsx`:

https://github.com/automattic/wp-calypso/blob/4537033c6794449ce0a10b69d72daf5fa658d553/client/login/magic-login/handle-emailed-link-form.jsx#L128

One approach might be to call `user.clear()` and then within the `then()` call, restore the data we wish to preserve.

## To dos

* [ ] Update `rebootAfterLogin` function in `client/login/magic-login/handle-emailed-link-form.jsx` to restore the localStorage that we need to hold on to for Gutenboarding
* [ ] Do we need to do the same for a logged in user who switches to another logged in user, or is that an edge case not worth pursuing? (This is just an idea off the top of my head from looking at [`client/blocks/login/index.jsx`](https://github.com/automattic/wp-calypso/blob/4537033c6794449ce0a10b69d72daf5fa658d553/client/blocks/login/index.jsx#L192)

**Comment by ramonjd** [comment]
@andrewserong You came up with a pretty nice solution back in the Zelda days :)
https://github.com/Automattic/wp-calypso/pull/34770/files#diff-3438d0da23b3176f5d05d0d0650da6a3R141

**Comment by andrewserong** [comment]
Just spent some more time with this issue this afternoon. I have a very small PR that will fix the issue for passwordless users / `Email me a login link` users who open the email up in the same browser context (https://github.com/Automattic/wp-calypso/pull/42319), however this change is effectively included in @p-jackson's refactor in https://github.com/Automattic/wp-calypso/pull/42318, anyway, so I'll abandon my PR in favour of the refactor.

While that solves part of the issue, something that will need a bit more work is to persist the Gutenboarding data past the login boundary for _any_ session, without requiring the localStorage state. For example, as a passwordless user / email me a login link user, you click the login link and open up a different browser context to the one you used initially, but it should still remember what you put into Gutenboarding. We already have a `plan` route fragment, so if we were to use a query string for the remaining data, we'd only need to add in, `domain`, `vertical`, `siteTitle`, `selectedDesign` and `selectedFonts`.

I think that's probably lower priority at this stage, so once #42318 lands, I think we could park the task for the time being, unless it seems more urgent to anyone?

**Comment by ramonjd** [comment]
> I think we could park the task for the time being, unless it seems more urgent to anyone?

Parked it in the backlog. If that makes sense, carry on!

**Comment by ramonjd** [comment]
Is this still relevant? I've a mind to close it since so much has changed since the good old days :)

**Comment by andrewserong** [comment]
Thanks for the reminder, Ramon! Agreed, I'll close it out since this issue is from so long ago, and the Gutenboarding flow is currently only in use (in practice) for logged in users.
